lib/spack/spack/directives.py

```diff
@@ -259,6 +262,29 @@ def variant(pkg, name, default=False, description=""):
     pkg.variants[name] = Variant(default, description)
 
 
+@directive('resources')
+def resource(pkg, **kwargs):
+    """
+    Define an external resource to be fetched and staged when building the package. Based on the keywords present in the
+    dictionary the appropriate FetchStrategy will be used for the resource.
+
+    List of recognized keywords:
+
+    * 'when' : represents the condition upon which the resource is needed (optional)
+    * 'destination' : path where to extract / checkout the resource (optional)
+
+    """
+    when = kwargs.get('when', pkg.name)
+    # FIXME : currently I assume destination to be a relative path (rooted at pkg.stage.source_path)
+    destination = kwargs.get('destination', "")
+    when_spec = parse_anonymous_spec(when, pkg.name)
+    resources = pkg.resources.setdefault(when_spec, [])
+    # FIXME : change URLFetchStrategy with a factory that selects based on kwargs
+    fetcher = URLFetchStrategy(**kwargs)
```

Comment:
There's already something like this in `fetch_strategy.py` -- it's the `for_package_version` method.  It's more complicated than you need here because it allows you to supply some default args in the actual package class, e.g.:

```python
    class Foo(Package):
        git = 'http://example.com/repo.git'

        version('2.6', tag='v2.6')
        version('2.7', tag='v2.7')
```

That should actually work and it'll instantiate git fetchers for both versions.  For resources, that's not necessary (or desired) so I think a factory would work well.  You could either do your static method, or you could add something like `from_args()` down where `from_url` and `for_package_version` are.

I think the only code you really need is this, from the middle of `for_package_version`:
```python
    for fetcher in all_strategies:                                                                                         
        if fetcher.matches(args):                                                                                          
            return fetcher(**args)   
```

I believe I had a simpler version of this in its own method before, but it wasn't used anywhere so I just left it out rather than having dead code.